src/main/java/org/pac4j/j2e/filter/RequiresAuthenticationFilter.java

```diff
@@ -23,59 +23,67 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+import org.pac4j.core.authorization.Authorizer;
+import org.pac4j.core.authorization.IsAuthenticatedAuthorizer;
+import org.pac4j.core.authorization.RequireAllRolesAuthorizer;
+import org.pac4j.core.authorization.RequireAnyRoleAuthorizer;
 import org.pac4j.core.client.Client;
+import org.pac4j.core.client.DirectClient;
 import org.pac4j.core.context.HttpConstants;
 import org.pac4j.core.context.J2EContext;
+import org.pac4j.core.context.Pac4jConstants;
 import org.pac4j.core.context.WebContext;
 import org.pac4j.core.credentials.Credentials;
 import org.pac4j.core.exception.RequiresHttpAction;
+import org.pac4j.core.exception.TechnicalException;
 import org.pac4j.core.profile.CommonProfile;
+import org.pac4j.core.profile.ProfileManager;
+import org.pac4j.core.profile.UserProfile;
 import org.pac4j.core.util.CommonHelper;
-import org.pac4j.j2e.configuration.ClientsConfiguration;
-import org.pac4j.j2e.util.UserUtils;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 /**
- * This filter aims to protect a secured resource.<br>
- * It handles both statefull (default) or stateless resources by delegating to a pac4j client.<br>
- *  - If statefull, it relies on the session and on the callback filter to terminate the authentication process.<br>
- *  - If stateless it validates the provided credentials and forward the request to
- * the underlying resource if the authentication succeeds.<br>
- * The filter also handles basic authorization based on two parameters: requireAnyRole and requireAllRoles.
+ * <p>This filter aims to protect a secured (stateful or stateless) resource.</p>
+ * <ul>
+ *  <li>If statefull, it relies on the session and on the callback filter to terminate the authentication process.</li>
+ *  <li>If stateless it validates the provided credentials and forward the request to the underlying resource if the authentication succeeds.</li>
+ * </ul>
+ * <p>The filter handles authorization based on an <code>Authorizer</code> (possibly built via two parameters: requireAnyRole and requireAllRoles).</p>
  * 
  * @author Jerome Leleu, Michael Remond
  * @since 1.0.0
  */
-@SuppressWarnings({ "unchecked", "rawtypes" })
+@SuppressWarnings("unchecked")
 public class RequiresAuthenticationFilter extends ClientsConfigFilter {
 
-    private static final Logger logger = LoggerFactory.getLogger(RequiresAuthenticationFilter.class);
+    protected String clientName;
 
-    private String clientName;
+    protected boolean isAjax = false;
 
-    private boolean stateless = false;
+    protected String requireAnyRole;
 
-    private boolean isAjax = false;
+    protected String requireAllRoles;
 
-    private String requireAnyRole;
+    protected Authorizer authorizer = new IsAuthenticatedAuthorizer();
 
-    private String requireAllRoles;
+    protected boolean useSessionForDirectClient = false;
```

Comment:
Two new concepts: `useSessionForDirectClient ` and `allowDynamicClientSelection `...